util/util.go

```diff
@@ -535,6 +536,121 @@ func (o MachinesByCreationTimestamp) Less(i, j int) bool {
+// or updates where Cluster.Spec.Paused transitions from true to false.
+// This implements a common requirement for Cluster infrastructure providers
+func WatchOnClusterUnPaused(c controller.Controller, mapFunc handler.Mapper, logger logr.Logger) error {
+	return c.Watch(
+		&source.Kind{Type: &clusterv1.Cluster{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: mapFunc,
+		},
+		predicate.Funcs{
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				oldCluster := e.ObjectOld.(*clusterv1.Cluster)
+				newCluster := e.ObjectNew.(*clusterv1.Cluster)
+				log := logger.WithValues("predicate", "updateEvent", "namespace", newCluster.Namespace, "cluster", newCluster.Name)
+				switch {
+				// return true if Cluster.Spec.Paused has changed from true to false
+				case oldCluster.Spec.Paused && !newCluster.Spec.Paused:
+					log.V(4).Info("Cluster was unpaused, will attempt to map associated resource.")
+					return true
+				// otherwise, return false
+				default:
+					log.V(4).Info("Cluster did not match expected conditions, will not attempt to map associated resource.")
+					return false
+				}
+			},
+			CreateFunc: func(e event.CreateEvent) bool {
+				cluster := e.Object.(*clusterv1.Cluster)
+				log := logger.WithValues("predicate", "createEvent", "namespace", cluster.Namespace, "cluster", cluster.Name)
+
+				// Only need to trigger a reconcile if the Cluster.Spec.Paused is false
+				if !cluster.Spec.Paused {
+					log.V(4).Info("Cluster is not paused, will attempt to map associated resource.")
+					return true
+				}
+				log.V(4).Info("Cluster did not match expected conditions, will not attempt to map associated resource.")
+				return false
+			},
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				log := logger.WithValues("predicate", "deleteEvent", "namespace", e.Meta.GetNamespace(), "cluster", e.Meta.GetName())
+				log.V(4).Info("Cluster did not match expected conditions, will not attempt to map associated resource.")
+				return false
+			},
+			GenericFunc: func(e event.GenericEvent) bool {
+				log := logger.WithValues("predicate", "genericEvent", "namespace", e.Meta.GetNamespace(), "cluster", e.Meta.GetName())
+				log.V(4).Info("Cluster did not match expected conditions, will not attempt to map associated resource.")
+				return false
+			},
+		},
+	)
+}
+
+// WatchOnClusterUnPausedAndInfrastructureReady adds a conditional watch to the controller given as input
+// that sends watch notifications on any create event where Cluster.Spec.Paused is false and
+// Cluster.Status.InfrastructureReady is true or updates where Cluster.Spec.Paused transitions from true to false
+// or updates where Cluster.Status.InfrastructureReady transitions from false to true
+// This implements a common requirement for Machine infrastructure providers
+func WatchOnClusterUnPausedAndInfrastructureReady(c controller.Controller, mapFunc handler.Mapper, logger logr.Logger) error {
```

Comment:
That thought also crossed my mind, but it has slightly different semantics (at least on Create) and could also potentially lead to double reconciliation of a resource if it happens to trigger both Watches, which I wanted to avoid.

Reply:
Would add some docs explaining why both exist and where an infrastructure provider should use them. Could get inadvertently removed later otherwise.

Reply:
Updated the godoc and examples, ptal